File: packages/cli/src/commands/episode.ts
import { Command } from "commander";
import { inspectEpisode, verifyEpisodeChain } from "../core/episode.js";
import { CliError } from "../core/exit.js";

export function episodeCommand(): Command {
  const episode = new Command("episode").description(
    "Inspect and verify x-harness episode packages"
  );

  episode
    .command("inspect")
    .description("Inspect an episode directory or .tar.gz bundle")
    .argument("<path>", "Episode directory or bundle path")
    .option("--json", "Output JSON instead of text", false)
    .action(async (episodePath: string, opts: { json?: boolean }) => {
      const result = await inspectEpisode(episodePath);
      if (opts.json) {
        console.log(JSON.stringify(result, null, 2));
      } else {
        console.log("# x-harness Episode Inspect");
        console.log(`- ok: ${result.ok}`);
        console.log(`- episode_id: ${result.episode_id ?? "unknown"}`);
        console.log(`- task_id: ${result.task_id ?? "unknown"}`);
        console.log(`- file_count: ${result.file_count}`);
        if (result.errors.length > 0) {
          console.log("");
          console.log("## Errors");
          for (const error of result.errors) console.log(`- ${error}`);
        }
        if (result.warnings.length > 0) {
          console.log("");
          console.log("## Warnings");
          for (const warning of result.warnings) console.log(`- ${warning}`);
        }
      }
      if (!result.ok) {
        throw new CliError("episode validation failed", 1);
      }
    });

  episode
    .command("verify-chain")
    .description("Verify episode chain integrity for a task")
    .requiredOption("--task-id <id>", "Task id")
    .option("--episodes-dir <dir>", "Episodes directory", ".x-harness/episodes")
    .option("--json", "Output JSON instead of text", false)
    .action(
      async (opts: {
        taskId: string;
        episodesDir?: string;
        json?: boolean;
      }) => {
        const result = await verifyEpisodeChain(opts.taskId, opts.episodesDir);
        if (opts.json) {
          console.log(JSON.stringify(result, null, 2));
        } else if (result.ok) {
          console.log(
            `episode chain valid: ${result.episodes_checked} episode(s) checked`
          );
          for (const id of result.episode_ids) console.log(`- ${id}`);
        } else {
          console.log("episode chain invalid:");
          for (const error of result.errors) console.log(`- ${error}`);
        }
        if (!result.ok) {
          throw new CliError("episode chain validation failed", 1);
        }
      }
    );

  episode
    .command("sign")
    .description("Signing skeleton for future release hardening")
    .option("--episode <path>", "Episode directory or bundle")
    .option("--mode <mode>", "Signing mode", "unsigned")
    .option("--json", "Output JSON instead of text", false)
    .action((opts: { episode?: string; mode?: string; json?: boolean }) => {
      const result = {
        implemented: false,
        mode: opts.mode ?? "unsigned",
        episode: opts.episode ?? null,
        note: "local MVP supports unsigned episode packages only; signing is reserved for release hardening",
      };
      if (opts.json) {
        console.log(JSON.stringify(result, null, 2));
      } else {
        console.log("episode sign is not implemented in local MVP.");
        console.log(result.note);
      }
    });

  return episode;
}

File: internal/cli/episode.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"github.com/BrianNguyen29/x-harness/internal/episode"
)

func handleEpisode(args []string, stdout io.Writer, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, "episode requires a subcommand: inspect, verify-chain")
		return ExitUsage
	}

	switch args[0] {
	case "inspect":
		return handleEpisodeInspect(args[1:], stdout, stderr)
	case "verify-chain":
		return handleEpisodeVerifyChain(args[1:], stdout, stderr)
	default:
		fmt.Fprintf(stderr, "unknown episode subcommand: %s\n", args[0])
		return ExitUsage
	}
}

func handleEpisodeInspect(args []string, stdout io.Writer, stderr io.Writer) int {
	jsonOutput := false
	var path string

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--json":
			jsonOutput = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			if path != "" {
				fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
				return ExitUsage
			}
			path = args[i]
		}
	}

	if path == "" {
		fmt.Fprintln(stderr, "Error: episode inspect requires a path argument")
		return ExitUsage
	}

	result, err := episode.InspectEpisode(path)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	if jsonOutput {
		data, err := json.MarshalIndent(result, "", "  ")
		if err != nil {
			fmt.Fprintf(stderr, "Error: failed to serialize result: %v\n", err)
			return ExitError
		}
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintln(stdout, "# x-harness Episode Inspect")
		fmt.Fprintf(stdout, "- ok: %v\n", result.OK)
		episodeID := "unknown"
		if result.EpisodeID != nil {
			episodeID = *result.EpisodeID
		}
		taskID := "unknown"
		if result.TaskID != nil {
			taskID = *result.TaskID
		}
		fmt.Fprintf(stdout, "- episode_id: %s\n", episodeID)
		fmt.Fprintf(stdout, "- task_id: %s\n", taskID)
		fmt.Fprintf(stdout, "- file_count: %d\n", result.FileCount)
		if len(result.Errors) > 0 {
			fmt.Fprintln(stdout, "")
			fmt.Fprintln(stdout, "## Errors")
			for _, error := range result.Errors {
				fmt.Fprintf(stdout, "- %s\n", error)
			}
		}
		if len(result.Warnings) > 0 {
			fmt.Fprintln(stdout, "")
			fmt.Fprintln(stdout, "## Warnings")
			for _, warning := range result.Warnings {
				fmt.Fprintf(stdout, "- %s\n", warning)
			}
		}
	}

	if !result.OK {
		return ExitError
	}
	return ExitOK
}

func handleEpisodeVerifyChain(args []string, stdout io.Writer, stderr io.Writer) int {
	taskID := ""
	episodesDir := ".x-harness/episodes"
	jsonOutput := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--task-id":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "Error: --task-id requires a value")
				return ExitUsage
			}
			i++
			taskID = args[i]
		case "--episodes-dir":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "Error: --episodes-dir requires a value")
				return ExitUsage
			}
			i++
			episodesDir = args[i]
		case "--json":
			jsonOutput = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if taskID == "" {
		fmt.Fprintln(stderr, "Error: --task-id is required")
		return ExitUsage
	}

	result, err := episode.VerifyEpisodeChain(taskID, episodesDir)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	if jsonOutput {
		data, err := json.MarshalIndent(result, "", "  ")
		if err != nil {
			fmt.Fprintf(stderr, "Error: failed to serialize result: %v\n", err)
			return ExitError
		}
		fmt.Fprintln(stdout, string(data))
	} else if result.OK {
		fmt.Fprintf(stdout, "episode chain valid: %d episode(s) checked\n", result.EpisodesChecked)
		for _, id := range result.EpisodeIDs {
			fmt.Fprintf(stdout, "- %s\n", id)
		}
	} else {
		fmt.Fprintln(stdout, "episode chain invalid:")
		for _, e := range result.Errors {
			fmt.Fprintf(stdout, "- %s\n", e)
		}
	}

	if !result.OK {
		return ExitError
	}
	return ExitOK
}
